### PR TITLE
[DROOLS-4351][DROOLS-4353] refactor CEP remote API + implement fireUntilHalt and fireAllRules

### PR DESCRIPTION
**Pull request by mariofusco**

### CODE REVIEW

File: hacep/src/main/java/org/kie/hacep/consumer/CommandHandler.java [hunk 1]
```diff
@@ -50,37 +52,53 @@ public CommandHandler(KieSessionContext kieSessionContext,
...
+        producer.produceSync(config.getKieSessionInfosTopicName(), command.getId(), fires);
+    }
+
     @Override
     public void visit(InsertCommand command) {
         RemoteFactHandle remoteFH = command.getFactHandle();
         FactHandle fh = kieSessionContext.getKieSession().getEntryPoint(command.getEntryPoint()).insert(remoteFH.getObject());
         kieSessionContext.getFhManager().registerHandle(remoteFH, fh);
-        kieSessionContext.getKieSession().fireAllRules();
+        if (command.isAutoFire()) {
```
Comment (danielezonca) [inline hacep/src/main/java/org/kie/hacep/consumer/CommandHandler.java]: I think it is wrong to have `isAutoFire` as property of the command. This means that status (fire or wait) is stored on client side and breaks in case of multiple clients.
I think this property should be stored inside drools executor and also inside snapshot status.
Wdyt?

File: hacep/src/main/java/org/kie/hacep/consumer/FactHandlesManager.java [hunk 2]
```diff
@@ -71,7 +72,7 @@ public FactHandle mapRemoteFactHandle(RemoteFactHandle remoteFH) {
 
     private InternalFactHandle getFactHandleById(RemoteFactHandle remoteFH) {
         int id = fhIdMap.get(remoteFH);
-        for (FactHandle fh : kieSession.getFactHandles()) {
+        for (FactHandle fh : kieSession.getFactHandles( new ClassObjectFilter( remoteFH.getObject().getClass() ) )) {
```
Comment (danielezonca) [inline hacep/src/main/java/org/kie/hacep/consumer/FactHandlesManager.java]: Good catch, do we have a test that cover this?

File: integration/src/test/java/org/kie/hacep/RemoteKieSessionImplTest.java [hunk 3]
```diff
@@ -68,6 +68,7 @@ public void getFactCountTest() {
                                                     RemoteKieSession.class);
         try (RemoteKieSessionImpl client = new RemoteKieSessionImpl(Config.getProducerConfig("getFactCountTest"),
                                                                     topicsConfig)) {
+            client.fireUntilHalt();
             client.listen();
```
Comment (danielezonca) [inline integration/src/test/java/org/kie/hacep/RemoteKieSessionImplTest.java]: This is not part of this PR but can you please verify what is `listen()`?
I do not expect user need to explicitly start the client. In any case, if we keep this command I think it needs to be the first method to be invoked

File: kie-remote/src/main/java/org/kie/remote/RemoteStatefulSession.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2005 JBoss Inc
```
Comment (danielezonca) [inline kie-remote/src/main/java/org/kie/remote/RemoteStatefulSession.java]: Please update year and company :)

File: kie-remote/src/main/java/org/kie/remote/RemoteWorkingMemory.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2005 JBoss Inc
```
Comment (danielezonca) [inline kie-remote/src/main/java/org/kie/remote/RemoteWorkingMemory.java]: Please update year and company :)

File: kie-remote/src/main/java/org/kie/remote/impl/AbstractRemoteEntryPoint.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2005 JBoss Inc
```
Comment (danielezonca) [inline kie-remote/src/main/java/org/kie/remote/impl/AbstractRemoteEntryPoint.java]: Please update year and company :)

File: kie-remote/src/main/java/org/kie/remote/impl/RemoteStatefulSessionImpl.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2005 JBoss Inc
```
Comment (danielezonca) [inline kie-remote/src/main/java/org/kie/remote/impl/RemoteStatefulSessionImpl.java]: Please update year and company :)

File: kie-remote/src/main/java/org/kie/remote/RemoteStatefulSession.java (hunk 4 above)
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/RemoteStatefulSession.java]: Why 2005 and JBoss?

File: kie-remote/src/main/java/org/kie/remote/RemoteWorkingMemory.java (hunk 5 above)
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/RemoteWorkingMemory.java]: Why 2005 and JBoss ?

File: kie-remote/src/main/java/org/kie/remote/command/EventInsertCommand.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.remote.command;
+
+import java.io.Serializable;
+
+public class EventInsertCommand extends WorkingMemoryActionCommand implements VisitableCommand,
```
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/command/EventInsertCommand.java]: InsertCommand and EventInsertCommand have different purpose ? The names are too similar

File: kie-remote/src/main/java/org/kie/remote/impl/AbstractRemoteEntryPoint.java (hunk 6 above)
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/impl/AbstractRemoteEntryPoint.java]: Why 2005 and JBoss ?

File: kie-remote/src/main/java/org/kie/remote/impl/RemoteKieSessionImpl.java [hunk 9]
```diff
@@ -40,6 +46,21 @@ public void close() {
 
     @Override
     public RemoteEntryPoint getEntryPoint( String name ) {
-        return getEntryPoint(name);
+        return entryPoints.computeIfAbsent( name, k -> new RemoteEntryPointImpl(sender, k, topicsConfig, requestsStore, delegate) );
```
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/impl/RemoteKieSessionImpl.java]: If this could be a concurrent hash map in the future, please note that Java 8 had two concurrency bugs in ConcurrentHashMap's computeIfAbsent().

File: kie-remote/src/main/java/org/kie/remote/impl/RemoteStatefulSessionImpl.java (hunk 7 above)
Comment (desmax74) [inline kie-remote/src/main/java/org/kie/remote/impl/RemoteStatefulSessionImpl.java]: Why 2005 and JBoss ?
